### PR TITLE
Masternodes must have required services enabled

### PR DESCRIPTION
**Pull request by UdjinM6**

### CODE REVIEW

**Comment by akshaynexus** [comment]
Won't this require mns to reindex for those mns who currently don't have txindex enabled?

**Comment by UdjinM6** [comment]
@akshaynexus correct. Such masternodes should not exist.

**Comment by akshaynexus** [comment]
https://docs.dash.org/en/stable/masternodes/setup.html

The documentation doesn't make a mention of txindex being required,might eb a good edit /addition to that if this pr is merged.

**Comment by UdjinM6** [comment]
@akshaynexus indeed, there are no explicit requirements regarding config options in docs as far as I can see. Might be a good idea to mention some @thephez ?

File: src/evo/mnauth.cpp [hunk 1]
```diff
@@ -67,6 +67,12 @@ void CMNAuth::ProcessMessage(CNode* pnode, const std::string& strCommand, CDataS
             return;
         }
 
+        if ((~pnode->nServices) & (NODE_NETWORK | NODE_BLOOM)) {
```
Comment (PastaPastaPasta) [inline src/evo/mnauth.cpp]: Can you add a comment here? It's not quite instantly obvious what's happening

File: src/init.cpp [hunk 2]
```diff
@@ -911,6 +911,15 @@ void InitParameterInteraction()
         // masternodes MUST accept connections from outside
         gArgs.ForceSetArg("-listen", "1");
         LogPrintf("%s: parameter interaction: -masternodeblsprivkey=... -> setting -listen=1\n", __func__);
+        // masternodes MUST have transaction index enabled
+        gArgs.ForceSetArg("-txindex", "1");
```
Comment (PastaPastaPasta) [inline src/init.cpp]: Isn't txindex on by default? Should this log message only occur if they try to disable txindex on a MN?

File: src/init.cpp [hunk 3]
```diff
@@ -911,6 +911,15 @@ void InitParameterInteraction()
...
+        LogPrintf("%s: parameter interaction: -masternodeblsprivkey set -> setting -txindex=1\n", __func__);
+        // masternodes MUST have bloom filters enabled
+        gArgs.ForceSetArg("-peerbloomfilters", "1");
+        LogPrintf("%s: parameter interaction: -masternodeblsprivkey set -> setting -peerbloomfilters=1\n", __func__);
+        // masternodes MUST NOT have any pruning enabled
+        gArgs.ForceSetArg("-prune", "0");
+        LogPrintf("%s: parameter interaction: -masternodeblsprivkey set -> setting -prune=0\n", __func__);
```
Comment (PastaPastaPasta) [inline src/init.cpp]: same

File: src/init.cpp [hunk 4]
```diff
@@ -911,6 +911,15 @@ void InitParameterInteraction()
         // masternodes MUST accept connections from outside
         gArgs.ForceSetArg("-listen", "1");
         LogPrintf("%s: parameter interaction: -masternodeblsprivkey=... -> setting -listen=1\n", __func__);
+        // masternodes MUST have transaction index enabled
+        gArgs.ForceSetArg("-txindex", "1");
+        LogPrintf("%s: parameter interaction: -masternodeblsprivkey set -> setting -txindex=1\n", __func__);
+        // masternodes MUST have bloom filters enabled
+        gArgs.ForceSetArg("-peerbloomfilters", "1");
+        LogPrintf("%s: parameter interaction: -masternodeblsprivkey set -> setting -peerbloomfilters=1\n", __func__);
```
Comment (PastaPastaPasta) [inline src/init.cpp]: same

**Comment by strophy** [comment]
I have set up a new branch in the docs and [added a note](https://github.com/dashpay/docs/commit/71997f2d79c09c9d16b5436eaf6cdb78d7319f26) that these flags will be automatically enabled once masternode mode is enabled.

File: src/init.cpp (hunk 2 above)
Comment (UdjinM6) [reply src/init.cpp]: Good point. In fact our code here was already doing the same for `listen` for example which is not quite right. I refactored the whole thing, pls see 6ead5e3

File: src/init.cpp [hunk 5]
```diff
@@ -992,6 +975,13 @@ void InitParameterInteraction()
         gArgs.ForceSetArg("-checklevel", "4");
         LogPrintf("%s: parameter interaction: additional indexes -> setting -checklevel=4\n", __func__);
     }
+
+    if (gArgs.IsArgSet("-masternodeblsprivkey")) {
+        if (gArgs.SoftSetArg("-maxconnections", itostr(DEFAULT_MAX_PEER_CONNECTIONS))) {
```
Comment (PastaPastaPasta) [inline src/init.cpp]: isn't masternode max peer count supposed to be higher than normal peers?

File: src/init.cpp [hunk 6]
```diff
@@ -1448,6 +1438,27 @@ bool AppInitParameterInteraction()
         InitWarning(_("-masternode option is deprecated and ignored, specifying -masternodeblsprivkey is enough to start this node as a masternode."));
     }
 
+    if (gArgs.IsArgSet("-masternodeblsprivkey")) {
+        if (!gArgs.GetBoolArg("-listen", DEFAULT_LISTEN)) {
```
Comment (PastaPastaPasta) [inline src/init.cpp]: this seems error prone in case we ever change these defaults.

Also, my thought process was that we could wrap 
```
        gArgs.ForceSetArg("-listen", "1");
        LogPrintf("%s: parameter interaction: -masternodeblsprivkey=... -> setting -listen=1\n", __func__);
```
in a `if gArgs.GetArg("-listen", 1)`

File: src/init.cpp (hunk 5 above)
Comment (UdjinM6) [reply src/init.cpp]: No, but it supposed to be _at least_ `DEFAULT_MAX_PEER_CONNECTIONS` while normal nodes might reduce it. Anyway, this code is actually a redundant one 🙈 removed it.

File: src/init.cpp (hunk 6 above)
Comment (UdjinM6) [reply src/init.cpp]: This specific part doesn't actually depend on the default value in any way. If we ever change defaults we should consider adding corresponding `SoftSetArg` in `InitParameterInteraction` instead.

Comment (PastaPastaPasta) [reply src/init.cpp]: I don't understand how this doesn't depend on the default value

Comment (UdjinM6) [reply src/init.cpp]: Let me rephrase it. This reads `-listen` from config/cmd-line options and _if none was found_ assigns the value we specified (in this case `DEFAULT_LISTEN`). But in any case, we expect it to be `false` here regardless of the (default) value. If we ever change defaults to `false`, we should add
```
if (gArgs.IsArgSet("-masternodeblsprivkey")) {
    if (gArgs.SoftSetBoolArg("-listen", true)) {
        LogPrintf(...);
    }
}
```
in `InitParameterInteraction`. But this `if (!gArgs.GetBoolArg("-listen", DEFAULT_LISTEN)) {` code here should be kept untouched.

Comment (PastaPastaPasta) [reply src/init.cpp]: if `-listen` hasn't been assigned, wouldn't it already be `DEFAULT_LISTEN`?

Comment (UdjinM6) [reply src/init.cpp]: I probably don't quite understand the issue I guess... Yes, if `-listen` hasn't been assigned yet, `GetBoolArg` should return `DEFAULT_LISTEN`. But it doesn't matter what the value of `DEFAULT_LISTEN` is, the code here is going to be the same both for `DEFAULT_LISTEN = true` and for `DEFAULT_LISTEN = false`. It's that code in `InitParameterInteraction` I mentioned above that should appear if `DEFAULT_LISTEN` would not match the expected result of `GetBoolArg` here.

Comment (PastaPastaPasta) [reply src/init.cpp]: Looking at it again, I think this is good.
